test/abc/sound/SequencePlayerTest.java

```diff
@@ -1,38 +1,125 @@
 package abc.sound;
 
-import static org.antlr.v4.runtime.CharStreams.fromFileName;
+import javax.sound.midi.InvalidMidiDataException;
+import javax.sound.midi.MidiUnavailableException;
 
-import java.io.IOException;
-
-import org.antlr.v4.runtime.CharStream;
-import org.antlr.v4.runtime.CommonTokenStream;
-import org.antlr.v4.runtime.tree.ParseTree;
-import org.antlr.v4.runtime.tree.ParseTreeWalker;
 import org.junit.Test;
 
-import abc.parser.AbcBaseListener;
-import abc.parser.AbcLexer;
-import abc.parser.AbcParser;
-
 public class SequencePlayerTest {
 
 	// TODO: warmup #2
 
 	@Test
-	public void test() {
-		String fileName = "sample_abc/piece1.abc";
+	public void piece1() {
+		try {
+			SequencePlayer player = new SequencePlayer(140, 12);
+
+			player.addNote(new Pitch('C').toMidiNote(), 0, 12);
+			player.addNote(new Pitch('C').toMidiNote(), 12, 12);
+			player.addNote(new Pitch('C').toMidiNote(), 24, 9);
+			player.addNote(new Pitch('D').toMidiNote(), 33, 3);
+			player.addNote(new Pitch('E').toMidiNote(), 36, 12);
+
+			player.addNote(new Pitch('E').toMidiNote(), 48, 9);
+			player.addNote(new Pitch('D').toMidiNote(), 57, 3);
+			player.addNote(new Pitch('E').toMidiNote(), 60, 9);
+			player.addNote(new Pitch('F').toMidiNote(), 69, 3);
+			player.addNote(new Pitch('G').toMidiNote(), 72, 24);
+
+			player.addNote(new Pitch('D').transpose(12).toMidiNote(), 96, 4);
+			player.addNote(new Pitch('D').transpose(12).toMidiNote(), 100, 4);
+			player.addNote(new Pitch('D').transpose(12).toMidiNote(), 104, 4);
+			player.addNote(new Pitch('G').toMidiNote(), 108, 4);
+			player.addNote(new Pitch('G').toMidiNote(), 112, 4);
+			player.addNote(new Pitch('G').toMidiNote(), 116, 4);
+
+			player.addNote(new Pitch('E').toMidiNote(), 120, 4);
+			player.addNote(new Pitch('E').toMidiNote(), 124, 4);
+			player.addNote(new Pitch('E').toMidiNote(), 128, 4);
+			player.addNote(new Pitch('C').toMidiNote(), 132, 4);
+			player.addNote(new Pitch('C').toMidiNote(), 136, 4);
+			player.addNote(new Pitch('C').toMidiNote(), 140, 4);
+
+			player.addNote(new Pitch('G').toMidiNote(), 144, 9);
+			player.addNote(new Pitch('F').toMidiNote(), 153, 3);
+			player.addNote(new Pitch('E').toMidiNote(), 156, 9);
+			player.addNote(new Pitch('D').toMidiNote(), 165, 3);
+			player.addNote(new Pitch('C').toMidiNote(), 168, 24);
+
+			System.out.println(player);
+
+			player.play();
+			Thread.sleep(10000);
```

Comment:
I ran into the need for `Thread.sleep()` as well (otherwise the player gets cut short. I wonder if there is another way to extend the lifetime of the thread during the execution of `player.play()`.

Reply:
Okay I found a way, here's a screenshot.
<img width="1680" alt="Screenshot 2022-11-26 at 3 22 56 PM" src="https://user-images.githubusercontent.com/14932673/204111066-f32ec02c-723b-4603-92ab-838afe0033e2.png">

Reply:
That looks good. Thanks for sharing it. Do you think it actually extends the lifetime of the thread? But it doesn't make any difference when I implement it. Both has the same lifetime.

Reply:
The difference is the sleep is dynamic rather than a static amount of time.

Reply:
Gotcha